processor/src/main/java/org/mapstruct/ap/internal/model/LifecycleCallbackFactory.java

```diff
@@ -125,8 +141,9 @@ private LifecycleCallbackFactory() {
                 ParameterAssignmentUtil.getParameterAssignments( availableParams, callback.getParameters() );
 
             if ( parameterAssignments != null
-                && callback.matches( extractSourceTypes( parameterAssignments ), method.getResultType() ) ) {
-
+                && callback.matches( extractSourceTypes( parameterAssignments ), method.getResultType() )
+                && ( callback.getReturnType().isVoid() || callback.getReturnType().isTypeVar()
+                    || callback.getReturnType().isAssignableTo( method.getResultType() ) ) ) {
```

Comment:
I would extract the or checks to some method instead of doing everything in the `if`.